design/src/rfcs/rfc0039_behavior_major_versions.md

```diff
@@ -0,0 +1,89 @@
+
+Both legacy and new SDKs have the need to alter their SDKs default. Historically, this caused new customers on legacy SDKs to be subject to legacy defaults, even when a better alternative existed.
+
+For new SDKs, a GA cutline presents difficult choices around timeline and features that can’t be added later without altering behavior.
+
+Both of these use cases are addressed by Behavior Major Versions.
+
+<!-- The "Terminology" section is optional but is really useful for defining the technical terms you're using in the RFC -->
+Terminology
+-----------
+
+- **Some Term**: A definition for that term
+
+<!-- Explain how users will use this new feature and, if necessary, how this compares to the current user experience -->
+The user experience if this RFC is implemented
+----------------------------------------------
+
+In the current version of the SDK, users can construct clients without indicating any sort of behavior major version.
+Once this RFC is implemented, there will be three ways to set a behavior major version:
+
+1. In code via `aws_config::from_env_with_version(BehaviorMajorVersion::latest())` and `<service>::Config::builder().behavior_major_version(...)`
+2. By enabling `behavior-version-latest` in either `aws-config` (which brings back `from_env`) OR a specific generated SDK crate
+
+```toml
+# Cargo.toml
+[dependencies]
+aws-config = { version = "1", features = ["behavior-version-latest"] }
+# OR
+aws-sdk-s3 = { version = "1", features = ["behavior-version-latest"] }
+```
+
+If no `BehaviorMajorVersion` is set, the client will panic during construction.
+
+`BehaviorMajorVersion` is an opaque struct with initializers like `::latest()`, `::v2023_11_09()`. Downstream code can check the version by calling methods like `::supports_v1()`
+
+When new BMV are added, the previous version constructor will be marked as `deprecated`. This serves as a mechanism to alert customers that a new BMV exists to allow them to upgrade.
+
+How to actually implement this RFC
+----------------------------------
+
+In order to implement this feature, we need to create a `BehaviorMajorVersion` struct, add config options to `SdkConfig` and `aws-config`, and wire it throughout the stack.
+```rust
+/// Behavior major-version of the client
+///
+/// Over time, new best-practice behaviors are introduced. However, these behaviors might not be backwards
+/// compatible. For example, a change which introduces new default timeouts or a new retry-mode for
+/// all operations might be the ideal behavior but could break existing applications.
+#[derive(Debug, Clone)]
+pub struct BehaviorMajorVersion {
+    // currently there is only 1 MV so we don't actually need anything in here.
+}
+```
+
+We also will add `BehaviorVersionLatestRuntimePlugin`, a runtime plugin that is enabled when the `behavior-version-latest` feature is enabled in runtime crates.
```

Comment:
Seems to me that these BMV plugins should be aggregate plugins that delegate to a bunch of other default RPs. Then the default RPs would no longer be configured on the client, but rather, just a single BMV plugin. That will cut down a lot on `Arc<dyn>` thrash since the BMV can directly refer to the RP type rather than making it into a trait object.

Separately, are there any use cases for registering a normal or override-level RP with BMVs?